tests/app/template_folder/test_template_folder_rest.py

```diff
@@ -188,3 +188,160 @@ def test_delete_template_folder_fails_if_folder_contains_templates(admin_request
     }
 
     assert TemplateFolder.query.count() == 1
+
+
+@pytest.mark.parametrize('data', [
+    {},
+    {'templates': None, 'folders': []},
+    {'folders': []},
+    {'templates': [], 'folders': [None]},
+    {'templates': [], 'folders': ['not a uuid']},
+])
+def test_move_to_folder_validates_schema(data, admin_request, notify_db_session):
+    admin_request.post(
+        'template_folder.move_to_template_folder',
+        service_id=uuid.uuid4(),
+        target_template_folder_id=uuid.uuid4(),
+        _data=data,
+        _expected_status=400
+    )
+
+
+def test_move_to_folder_moves_folders_and_templates(admin_request, sample_service):
+    target_folder = create_template_folder(sample_service, name='target')
+    f1 = create_template_folder(sample_service, name='f1')
+    f2 = create_template_folder(sample_service, name='f2')
+
+    t1 = create_template(sample_service, template_name='t1', folder=f1)
+    t2 = create_template(sample_service, template_name='t2', folder=f1)
+    t3 = create_template(sample_service, template_name='t3', folder=f2)
+    t4 = create_template(sample_service, template_name='t4', folder=target_folder)
+
+    admin_request.post(
+        'template_folder.move_to_template_folder',
+        service_id=sample_service.id,
+        target_template_folder_id=target_folder.id,
+        _data={
+            'templates': [str(t1.id)],
+            'folders': [str(f1.id)]
+        },
+        _expected_status=204
+    )
+
+    assert target_folder.parent is None
+    assert f1.parent == target_folder
+    assert f2.parent is None  # unchanged
+
+    assert t1.folder == target_folder  # moved out of f1, even though f1 is also being moved
+    assert t2.folder == f1  # stays in f1, though f1 has moved
+    assert t3.folder == f2  # unchanged
+    assert t4.folder == target_folder  # unchanged
+
+    # versions are all unchanged
+    assert t1.version == 1
+    assert t2.version == 1
+    assert t3.version == 1
+    assert t4.version == 1
+
+
+def test_move_to_folder_moves_folders_and_templates_to_top_level_if_no_target(admin_request, sample_service):
+    f1 = create_template_folder(sample_service, name='f1')
```

Comment:
I don't understand: we  want to test moving this folder out from another folder and to the top level, right? But when we create it, isn't it already at the top level?